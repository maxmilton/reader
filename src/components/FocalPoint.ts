import "./FocalPoint.xcss";

import { create } from "stage1/fast";

const NON_WORD_RE = /\W/;

/**
 * Get the Optimal Recognition Point (OPR) index for a given word phrase.
 */
export function indexOfORP(word: string): number {
  let len = word.length;

  while (NON_WORD_RE.test(word[--len]));

  if (NON_WORD_RE.test(word[0])) len++;

  switch (++len) {
    case 0:
    case 1:
      return 0;
    case 2:
    case 3:
      return 1;
    default:
      return Math.trunc(len / 2) - 1;
  }
}

export type FocalPointComponent = HTMLSpanElement;

const view = create("span");
view.id = "fp";

export function FocalPoint(char: string): FocalPointComponent {
  // Don't clone; reuse the same node.
  view.textContent = char;
  return view;
}
