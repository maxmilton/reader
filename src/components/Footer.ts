import './Footer.xcss';

import { compile } from 'stage1/macro' assert { type: 'macro' };
import { h } from 'stage1/runtime';
// import { interpolate } from '../macros' assert { type: 'macro' };
import { decodeEntities, removeNbsp } from '../macros' assert { type: 'macro' };

export type FooterComponent = HTMLElement;

// const meta = compile(
//   `
//     <footer>
//       © <a href=https://maxmilton.com class="normal muted" target=_blank>Max Milton</a> ・ v${process
//         .env
//         .APP_RELEASE!} ・ <a href=https://github.com/maxmilton/reader/issues target=_blank>report bug</a>
//     </footer>
//   `,
//   { keepSpaces: true },
// );
// const meta = compile(
//   // FIXME: This is a convoluted workaround for a bug in the bun macro system,
//   // where it crashes when doing string literal template interpolation. See:
//   // https://github.com/oven-sh/bun/issues/3830
//   interpolate(
//     `
//       <footer>
//         © <a href=https://maxmilton.com class="normal muted" target=_blank>Max Milton</a> ・ v%%1%% ・ <a href=https://github.com/maxmilton/reader/issues target=_blank>report bug</a>
//       </footer>
//     `,
//     [process.env.APP_RELEASE!],
//   ),
//   { keepSpaces: true },
// );
const meta = compile(
  // FIXME: This is an alternative workaround for the bun macro bug. See: https://github.com/oven-sh/bun/issues/3830
  decodeEntities(`
    <footer>
      &#169;&nbsp;<a href=https://maxmilton.com class="normal muted" target=_blank>Max Milton</a>&nbsp;&#12539; v${process.env.APP_RELEASE} &#12539;&nbsp;<a href=https://github.com/maxmilton/reader/issues target=_blank>report bug</a>
    </footer>
  `),
  // FIXME: Passing objects is also currently broken. See: https://github.com/oven-sh/bun/issues/3832
  // { keepSpaces: true },
);

export function Footer(): FooterComponent {
  // FIXME: Remove the removeNbsp macro once bun issue #3832 is fixed.
  return h<FooterComponent>(removeNbsp(meta.html));
  // return h<FooterComponent>(meta.html);
}
