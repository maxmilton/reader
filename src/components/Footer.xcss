footer {
  #apply: .mv2, .muted, .fss, .tc;
}
