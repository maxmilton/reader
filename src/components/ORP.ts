import './ORP.xcss';

import { create } from 'stage1';

const NON_WORD_RE = /\W/;

/**
 * Get the Optimal Recognition Point (OPR) index for a given word phrase.
 */
export function indexOfORP(word: string): number {
  let len = word.length;

  while (NON_WORD_RE.test(word[--len]));

  if (NON_WORD_RE.test(word[0])) len++;

  switch (++len) {
    case 0:
    case 1:
      return 0;
    case 2:
    case 3:
      return 1;
    default:
      return Math.trunc(len / 2) - 1;
  }
}

export type ORPComponent = HTMLSpanElement;

const view = create('span');
view.id = 'orp';

/** Optimal recognition point. */
export function ORP(char: string): ORPComponent {
  // Don't clone; reuse the same node.
  view.textContent = char;
  return view;
}
