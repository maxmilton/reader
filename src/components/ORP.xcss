#orp {
  color: ${x.color.primary};
  font-variant: none; // ligatures break ORP highlighting
  text-shadow: 0 0 1rem ${x.color.background};
}
