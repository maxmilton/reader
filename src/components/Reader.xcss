#progress,
#focus,
#word {
  contain: strict;
}

#progress,
#bar {
  width: ${x.app.width};
  height: 4px;
}

#bar {
  background-color: ${x.fn.color(x.color.rose1).mix(x.fn.color(x.color.background), 0.5)};
  transform: translateX(-100%);
  transition: transform 0.3s ease-out;
  will-change: transform;
}

#controls {
  display: flex;
  align-items: center;
  margin: 0.5rem 0.5rem 0;
}

#play {
  width: 10rem;
  margin-left: 0.5rem;
}

#focus {
  position: absolute;
  top: 74px;
  left: ${Number.parseInt(x.app.width) / 2 - 2}px;
  width: 4px;
  height: 92px;
  background:
    linear-gradient(
      to bottom,
      ${x.color.rose1} 0%,
      ${x.color.background} 38%,
      ${x.color.background} 62%,
      ${x.color.rose1} 100%
    );
  opacity: 0;
  transition: opacity 0.3s;

  &.show {
    opacity: 1;
    transition: opacity 1s;
  }
}

#word {
  position: relative;
  left: 50%;
  display: flex;
  align-items: center;
  height: 134px;
  font-family: ${x.fontStack};
  font-size: 48px;
  white-space: nowrap;
  transform: translateX(-50%);
}

#summary {
  width: 100%;
  font-size: 26px;
  text-align: center;
  white-space: initial; // allow long error text to wrap
  transition: opacity 0.8s;

  & > em {
    font-size: 48px;
  }
}
