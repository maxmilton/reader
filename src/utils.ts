export async function getCurrentTab(): Promise<chrome.tabs.Tab> {
  const [tab] = await chrome.tabs.query({ active: true, currentWindow: true });
  return tab;
}

/**
 * Execute a function in the context of the current active tab.
 *
 * Note: The function will be serialised + injected into the tab's page and
 * then run as a content script so it must be self contained (a pure function;
 * no access or mutation outside its own closure scope).
 *
 * @param func - Function to run in page. Must be self-contained.
 * @param args - Arguments to pass to function. Must be JSON-serializable.
 *
 * @see https://developer.chrome.com/docs/extensions/reference/scripting/#method-executeScript
 * @see https://developer.chrome.com/docs/extensions/reference/scripting/#type-ScriptInjection
 * @see https://developer.chrome.com/en/blog/crx-scripting-api/#injecting-a-function-with-arguments
 */
export async function exec<T, A extends any[] = []>(
  func: (..._args: A) => T,
  args?: A,
): Promise<T> {
  const tab = await getCurrentTab();

  const [{ result }] = await chrome.scripting.executeScript({
    target: { tabId: tab.id! },
    func,
    args,
  });

  return result as T;
}

/**
 * Get the Optimal Recognition Point (OPR) index for a given word phrase.
 */
export function indexOfORP(word: string): number {
  let len = word.length;

  while (/\W/.test(word[--len]));

  if (/\W/.test(word[0])) len++;

  switch (++len) {
    case 0:
    case 1: {
      return 0;
    }
    case 2:
    case 3: {
      return 1;
    }
    default: {
      return Math.floor(len / 2) - 1;
    }
  }
}
