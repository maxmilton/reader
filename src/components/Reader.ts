import './Reader.xcss';

import { collect, h } from 'stage1';
import { compile } from 'stage1/macro' assert { type: 'macro' };
import { extractText } from '../extractor';
import { exec } from '../utils';
import { ORP, indexOfORP } from './ORP';

// eslint-disable-next-line unicorn/prefer-top-level-await
const extractedWords = (async () => {
  performance.mark('Extract:Begin');
  const html = await exec(() => {
    const selection = window.getSelection();
    if (selection?.type === 'Range') {
      const range = selection.getRangeAt(0);
      const contents = range.cloneContents();
      const body = document.createElement('body');
      body.append(contents);
      return body.outerHTML;
    }
    return document.documentElement.outerHTML;
  });
  // eslint-disable-next-line prefer-template
  const text = ' 3. 2. 1. ' + extractText(html) + '\n';
  performance.measure('Extract', 'Extract:Begin');
  return text.split(' ');
})();

// https://github.com/cameron/squirt/blob/03cf7bf103652857bd54fa7960a39fc27e306b31/squirt.js#L168-L187
const WAIT_AFTER_SHORT_WORD = 1.2;
const WAIT_AFTER_LONG_WORD = 1.5;
const WAIT_AFTER_COMMA = 2;
const WAIT_AFTER_PERIOD = 3;
const WAIT_AFTER_PARAGRAPH = 3.5;

function waitMultiplier(word: string, forceWait?: boolean) {
  if (forceWait) return WAIT_AFTER_PERIOD;
  // if (word === 'Dr.' || word === 'Mr.' || word === 'Mrs.' || word === 'Ms.') return 1;

  // eslint-disable-next-line unicorn/prefer-at
  let lastChar = word[word.length - 1];
  // eslint-disable-next-line unicorn/prefer-at
  if (/["”]/.test(lastChar)) lastChar = word[word.length - 2];

  if (lastChar === '\n') return WAIT_AFTER_PARAGRAPH;
  if ('.!?…'.includes(lastChar)) return WAIT_AFTER_PERIOD;
  if (',;:–'.includes(lastChar)) return WAIT_AFTER_COMMA;
  if (word.length < 4) return WAIT_AFTER_SHORT_WORD;
  if (word.length > 11) return WAIT_AFTER_LONG_WORD;
  return 1;
}

export interface UserSettings {
  /** Target words per minute. */
  wpm?: number;
}

type ReaderComponent = HTMLDivElement;

type Refs = {
  progress: HTMLDivElement;
  rewind: HTMLButtonElement;
  play: HTMLButtonElement;
  slower: HTMLButtonElement;
  speed: Text;
  faster: HTMLButtonElement;
  focus: HTMLDivElement;
  w: HTMLDivElement;
};

const meta = compile(`
  <div>
    <div id=progress>
      <div @progress id=bar></div>
    </div>

    <div id=controls>
      <button @rewind>Rewind</button>
      <button @play id=play>Play</button>
      @speed
      <button @slower>−</button>
      <button @faster>+</button>
    </div>

    <div @focus id=focus></div>
    <div @w id=word></div>
  </div>
`);
const view = h<ReaderComponent>(meta.html);

export function Reader(): ReaderComponent {
  const root = view;
  const refs = collect<Refs>(root, meta.k, meta.d);
  const wordref = refs.w;
  let words: string[] = [];
  let wordsIndex = 0;
  /** Current target words per minute. */
  let wpm = 0;
  /** Current timing base rate; ms per word. */
  let rate = 0;
  let startTime = 0;
  let timer: number | undefined;

  function stop() {
    clearTimeout(timer);
    timer = undefined;
    refs.play.textContent = 'Play';
  }

  function end() {
    const time = (Date.now() - startTime) / 1000;
    stop();

    wordref.innerHTML = `<div id=summary><em>ﬁn.</em><br>You read ${
      // exclude intro countdown
      words.length - 4
    } words in ${
      time < 60
        ? `${Math.trunc(time)} seconds`
        : `${Math.trunc(time / 60)} minute${time < 120 ? '' : 's'}`
    }.</div>`;

    wordref.style.cssText = '';
    refs.progress.style.transform = 'translateX(0)';
    refs.focus.className = '';
    refs.play.textContent = 'Play again';
    wordsIndex = 0;
    startTime = 0;
  }

  function next(forceWait?: boolean) {
    if (++wordsIndex >= words.length) {
      end();
      return;
    }

    const word = words[wordsIndex];
    const orpIndex = indexOfORP(word);
    let orp;

    wordref.replaceChildren(
      word.slice(0, orpIndex),
      (orp = ORP(word[orpIndex])),
      word.slice(orpIndex + 1),
    );

    wordref.style.transform = `translateX(-${
      orp.offsetLeft + orp.offsetWidth / 2
    }px)`;
    refs.progress.style.transform = `translateX(${
      (wordsIndex / words.length - 1) * 100
    }%)`;

    timer = (setTimeout as Window['setTimeout'])(
      // eslint-disable-next-line @typescript-eslint/no-confusing-void-expression
      () => next(),
      rate * waitMultiplier(word, forceWait),
    );
  }

  function start(slowStart?: boolean) {
    if (!startTime) {
      startTime = Date.now();
    }

    next(slowStart);
    refs.focus.className = 'show';
    refs.play.textContent = 'Pause';
  }

  function updateWPM(newWPM: number) {
    rate = 60_000 / newWPM;
    refs.speed.nodeValue = `${newWPM} wpm`;
    refs.slower.disabled = newWPM <= 60;
    refs.faster.disabled = newWPM >= 1200;

    // Avoid writing to storage on initial load
    if (wpm) {
      void chrome.storage.sync.set({ wpm: newWPM });
    }

    wpm = newWPM;
  }

  refs.rewind.__click = () => {
    stop();

    // Go back 3 seconds worth of words
    wordsIndex -= 3000 / rate;

    if (wordsIndex <= 0) {
      wordsIndex = 0;
    } else {
      // Set index to start of the sentence
      while (wordsIndex-- && !/[!.?…]$/.test(words[wordsIndex]));
    }

    start(true);
  };

  refs.play.__click = () => {
    if (timer) {
      stop();
    } else {
      start();
    }
  };

  // eslint-disable-next-line @typescript-eslint/no-confusing-void-expression
  refs.slower.__click = () => updateWPM(wpm - 60);
  // eslint-disable-next-line @typescript-eslint/no-confusing-void-expression
  refs.faster.__click = () => updateWPM(wpm + 60);

  chrome.storage.sync
    .get()
    .then((settings: UserSettings) => {
      updateWPM(settings.wpm ?? 180);
      return extractedWords;
    })
    .then((wordList) => {
      performance.measure('Preprocessing');
      words = wordList;
      start(true);
    })
    .catch((error) => {
      wordref.innerHTML = `<div id=summary>${String(error)}</div>`;
      refs.rewind.disabled = true;
      refs.play.disabled = true;
      // eslint-disable-next-line no-console
      console.error(error);
    });

  return root;
}
