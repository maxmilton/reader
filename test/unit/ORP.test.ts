import { afterEach, expect, test } from 'bun:test';
import { ORP, type ORPComponent } from '../../src/components/ORP';
import { cleanup, render } from './utils';

afterEach(cleanup);

test('rendered DOM contains expected elements', () => {
  expect.assertions(4);
  const rendered = render(ORP('x'));
  const root = rendered.container.firstChild as HTMLSpanElement;
  expect(root).toBeTruthy();
  expect(rendered.container.querySelector('#orp')).toBe(root);
  expect(root).toBeInstanceOf(window.HTMLSpanElement);
  expect(root.textContent).toBe('x');
});

test('rendered DOM matches snapshot', () => {
  expect.assertions(1);
  const rendered = render(ORP('x'));
  expect(rendered.container.innerHTML).toMatchSnapshot();
});

test('returns the same reused node every call', () => {
  expect.assertions(10);
  let node1: ORPComponent;
  let node2: ORPComponent;
  let node3: ORPComponent;
  const rendered1 = render((node1 = ORP('1')));
  expect(rendered1.container.firstChild).toBeTruthy();
  const rendered2 = render((node2 = ORP('2')));
  // because it's the same node, it should have moved from the previous parent
  expect(rendered1.container.firstChild).toBe(null);
  expect(rendered2.container.firstChild).toBeTruthy();
  const rendered3 = render((node3 = ORP('3')));
  expect(rendered1.container.firstChild).toBe(null);
  expect(rendered2.container.firstChild).toBe(null);
  expect(rendered3.container.firstChild).toBeTruthy();
  expect(node1.nodeName).toBe('SPAN');
  expect(node1).toBe(node2);
  expect(node2).toBe(node3);
  expect(node3).toBe(node1);
});
