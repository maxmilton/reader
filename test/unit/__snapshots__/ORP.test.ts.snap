// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`rendered DOM matches snapshot 1`] = `"<span id="orp">x</span>"`;

exports[`rendered DOM matches snapshot 2`] = `"<span id="orp">x</span>"`;
